File: app/src/main/java/com/liadpaz/amp/fragments/PlaylistFragment.java
package com.liadpaz.amp.fragments;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.liadpaz.amp.R;
import com.liadpaz.amp.adapters.PlaylistAdapter;
import com.liadpaz.amp.adapters.SongsListAdapter;
import com.liadpaz.amp.databinding.FragmentPlaylistBinding;
import com.liadpaz.amp.dialogs.PlaylistsDialog;
import com.liadpaz.amp.interfaces.ItemTouchHelperAdapter;
import com.liadpaz.amp.interfaces.OnRecyclerItemClickListener;
import com.liadpaz.amp.livedatautils.PlaylistsUtil;
import com.liadpaz.amp.livedatautils.QueueUtil;
import com.liadpaz.amp.service.ServiceConnector;
import com.liadpaz.amp.utils.AmpPlaybackPreparer;
import com.liadpaz.amp.viewmodels.Playlist;
import com.liadpaz.amp.viewmodels.Song;

import java.util.ArrayList;
import java.util.Collections;

public class PlaylistFragment extends Fragment {
    private static final String TAG = "AmpApp.PlaylistFragment";

    private ListAdapter<Song, ? extends RecyclerView.ViewHolder> adapter;

    private Playlist playlist;

    private FragmentPlaylistBinding binding;

    private PlaylistFragment(Playlist playlist) { this.playlist = playlist; }

    @NonNull
    public static PlaylistFragment newInstance(Playlist playlist) { return new PlaylistFragment(playlist); }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return (binding = FragmentPlaylistBinding.inflate(inflater, container, false)).getRoot();
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        OnRecyclerItemClickListener onMoreClicked = (v, position) -> new PopupMenu(requireContext(), v) {{
            inflate(playlist.name.equals(getString(R.string.playlist_recently_added)) ? R.menu.menu_playlist_recently : R.menu.menu_playlist);
            setOnMenuItemClickListener(item -> {
                switch (item.getItemId()) {
                    case R.id.menuPlayNext: {
                        AmpPlaybackPreparer.addToQueueNext(adapter.getCurrentList().get(position));
                        break;
                    }

                    case R.id.menuAddQueue: {
                        AmpPlaybackPreparer.addToQueue(adapter.getCurrentList().get(position));
                        break;
                    }

                    case R.id.menuQueueAddPlaylist: {
                        new PlaylistsDialog(adapter.getCurrentList().get(position)).show(getChildFragmentManager(), null);
                        break;
                    }

                    case R.id.menuRemoveFromPlaylist: {
                        Playlist playlist = PlaylistsUtil.removePlaylist(PlaylistFragment.this.playlist.name);
                        playlist.songs.remove(position);
                        adapter.submitList(playlist.songs);
                        adapter.notifyItemRemoved(position);
                        PlaylistsUtil.addPlaylist(playlist);
                        break;
                    }
                }
                return true;
            });
        }}.show();
        View.OnClickListener onShuffleClickListener = v -> {
            if (playlist.songs.size() != 0) {
                ArrayList<Song> queue = new ArrayList<>(playlist.songs);
                Collections.shuffle(queue);
                QueueUtil.setQueue(queue);
                QueueUtil.setPosition(0);
                ServiceConnector.playFromQueue();
            }
        };

        if (playlist.name.equals(getString(R.string.playlist_recently_added))) {
            adapter = new SongsListAdapter(requireContext(), onMoreClicked, onShuffleClickListener);
            binding.btnDelete.setVisibility(View.GONE);
        } else {
            adapter = new PlaylistAdapter(requireContext(), onMoreClicked, onShuffleClickListener, new ItemTouchHelperAdapter() {
                @Override
                public void onItemMove(int fromPosition, int toPosition) {
                    Playlist playlist = PlaylistsUtil.removePlaylist(PlaylistFragment.this.playlist.name);
                    Collections.swap(playlist.songs, fromPosition, toPosition);
                    PlaylistsUtil.addPlaylist(playlist);
                }

                @Override
                public void onItemDismiss(int position) {}
            });

            new ItemTouchHelper(new ItemTouchHelper.Callback() {
                @Override
                public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
                    return makeMovementFlags(viewHolder.getAdapterPosition() == 0 ? 0 : ItemTouchHelper.DOWN | ItemTouchHelper.UP, 0);
                }

                @Override
                public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                    int toPosition = target.getAdapterPosition() == 0 ? 1 : target.getAdapterPosition();
                    ((ItemTouchHelperAdapter)adapter).onItemMove(viewHolder.getAdapterPosition(), toPosition);
                    return true;
                }

                @Override
                public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {}

                @Override
                public void onSelectedChanged(@Nullable RecyclerView.ViewHolder viewHolder, int actionState) {
                    if (actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
                        viewHolder.itemView.setBackgroundColor(Color.BLACK);
                    }
                }

                @Override
                public void clearView(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
                    viewHolder.itemView.setBackground(null);
                }
            }) {{
                attachToRecyclerView(binding.rvSongs);
                ((PlaylistAdapter)adapter).setOnStartDragListener(this::startDrag);
            }};

            binding.btnDelete.setOnClickListener(v -> new MaterialAlertDialogBuilder(requireContext()).setTitle(R.string.dialog_delete_playlist_title).setMessage(R.string.dialog_delete_playlist_content).setPositiveButton(R.string.dialog_yes, ((dialog, which) -> {
                PlaylistsUtil.removePlaylist(playlist.name);
                getParentFragmentManager().popBackStack();
            })).setNegativeButton(R.string.dialog_no, null).show());

            PlaylistsUtil.observe(requireActivity(), playlists -> {
                if (!PlaylistsUtil.getIsChanging()) {
                    for (Playlist playlist : playlists) {
                        if (playlist.name.equals(this.playlist.name)) {
                            adapter.submitList(playlist.songs);
                            return;
                        }
                    }
                } else {
                    PlaylistsUtil.setIsChanging(false);
                }
            });
        }

        binding.tvPlaylistName.setText(playlist.name);
        binding.rvSongs.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.rvSongs.addItemDecoration(new DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL));
        binding.rvSongs.setAdapter(adapter);

        adapter.submitList(playlist.songs);
    }
}

File: app/src/main/java/com/liadpaz/amp/fragments/PlaylistFragment.kt
package com.liadpaz.amp.fragments

import android.content.DialogInterface
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.*
import androidx.recyclerview.widget.ItemTouchHelper.*
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.liadpaz.amp.R
import com.liadpaz.amp.adapters.PlaylistAdapter
import com.liadpaz.amp.adapters.SongsListAdapter
import com.liadpaz.amp.databinding.FragmentPlaylistBinding
import com.liadpaz.amp.dialogs.PlaylistsDialog
import com.liadpaz.amp.interfaces.ItemTouchHelperAdapter
import com.liadpaz.amp.interfaces.OnRecyclerItemClickListener
import com.liadpaz.amp.interfaces.OnStartDragListener
import com.liadpaz.amp.livedatautils.PlaylistsUtil
import com.liadpaz.amp.livedatautils.QueueUtil
import com.liadpaz.amp.viewmodels.Playlist
import com.liadpaz.amp.viewmodels.Song
import java.util.*

class PlaylistFragment private constructor(private val playlist: Playlist?) : Fragment() {
    private val TAG = "AmpApp.PlaylistFragment"

    private lateinit var adapter: ListAdapter<Song, out RecyclerView.ViewHolder>

    private lateinit var binding: FragmentPlaylistBinding

    override fun onCreateView(inflater: LayoutInflater, container:
    ViewGroup?, savedInstanceState: Bundle?): View? {
        return FragmentPlaylistBinding.inflate(inflater, container, false).also { binding = it }.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val onMoreClicked = OnRecyclerItemClickListener { v: View?, position: Int ->
            PopupMenu(requireContext(), v).apply {
                inflate(if (playlist!!.name == getString(R.string.playlist_recently_added)) R.menu.menu_playlist_recently else R.menu.menu_playlist)
                setOnMenuItemClickListener { item: MenuItem ->
                    when (item.itemId) {
                        R.id.menuPlayNext -> {
//                                AmpPlaybackPreparer.addToQueueNext(adapter.currentList[position])
                            // TODO: play next
                        }
                        R.id.menuAddQueue -> {
//                                AmpPlaybackPreparer.addToQueue(adapter.currentList[position])
                            // TODO: add to queue
                        }
                        R.id.menuQueueAddPlaylist -> {
                            PlaylistsDialog(adapter.currentList[position]).show(childFragmentManager, null)
                        }
                        R.id.menuRemoveFromPlaylist -> {
                            val playlist = PlaylistsUtil.removePlaylist(playlist.name)
                            playlist!!.songs.removeAt(position)
                            adapter.notifyItemRemoved(position)
                            PlaylistsUtil.addPlaylist(playlist)
                        }
                    }
                    true
                }
                show()
            }
        }
        val onShuffleClickListener = View.OnClickListener {
            if (playlist!!.songs.size != 0) {
                val queue = ArrayList(playlist.songs)
                queue.shuffle()
                QueueUtil.queue.postValue(queue)
                QueueUtil.queuePosition.postValue(0)
                // TODO: play queue
            }
        }
        if (playlist!!.name == getString(R.string.playlist_recently_added)) {
            adapter = SongsListAdapter(requireContext(), onMoreClicked, onShuffleClickListener)
            binding.btnDelete.visibility = View.GONE
        } else {
            adapter = PlaylistAdapter(requireContext(), onMoreClicked, onShuffleClickListener, object : ItemTouchHelperAdapter {
                override fun onItemMove(fromPosition: Int, toPosition: Int) {
                    PlaylistsUtil.addPlaylist(PlaylistsUtil.removePlaylist(playlist.name).also {
                        Collections.swap(it!!.songs, fromPosition, toPosition)
                    }!!)
                }

                override fun onItemDismiss(position: Int) = Unit
            })
            ItemTouchHelper(object : ItemTouchHelper.Callback() {
                override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
                    return makeMovementFlags(if (viewHolder.adapterPosition == 0) 0 else DOWN or UP, 0)
                }

                override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                    val toPosition = if (target.adapterPosition == 0) 1 else target.adapterPosition
                    (adapter as ItemTouchHelperAdapter).onItemMove(viewHolder.adapterPosition, toPosition)
                    return true
                }

                override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}
                override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
                    if (actionState == ACTION_STATE_DRAG) {
                        viewHolder!!.itemView.setBackgroundColor(Color.BLACK)
                    }
                }

                override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
                    viewHolder.itemView.background = null
                }
            }).apply {
                attachToRecyclerView(binding.rvSongs)
                (adapter as PlaylistAdapter).setOnStartDragListener(OnStartDragListener{ viewHolder: RecyclerView.ViewHolder -> startDrag(viewHolder) })
            }
            binding.btnDelete.setOnClickListener {
                MaterialAlertDialogBuilder(requireContext()).setTitle(R.string.dialog_delete_playlist_title).setMessage(R.string.dialog_delete_playlist_content).setPositiveButton(R.string.dialog_yes) { _, _ ->
                    PlaylistsUtil.removePlaylist(playlist.name)
                    parentFragmentManager.popBackStack()
                }.setNegativeButton(R.string.dialog_no, null).show()
            }
            PlaylistsUtil.observe(requireActivity(), Observer { playlists: Queue<Playlist> ->
                if (!PlaylistsUtil.getIsChanging()) {
                    for (playlist in playlists) {
                        if (playlist.name == this.playlist.name) {
                            adapter.submitList(playlist.songs)
                            return@Observer
                        }
                    }
                } else {
                    PlaylistsUtil.setIsChanging(false)
                }
            })
        }
        binding.tvPlaylistName.text = playlist.name
        binding.rvSongs.layoutManager = LinearLayoutManager(requireContext())
        binding.rvSongs.addItemDecoration(DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL))
        binding.rvSongs.adapter = adapter
        adapter.submitList(playlist.songs)
    }

    companion object {
        fun newInstance(playlist: Playlist?): PlaylistFragment {
            return PlaylistFragment(playlist)
        }
    }
}